ocaml/src/fragment_cache_alba.ml

```diff
@@ -0,0 +1,163 @@
+                        ~manifest_cache_size
+                        ~bad_fragment_callback:(fun alba_client
+                                                    ~namespace_id
+                                                    ~object_id ~object_name
+                                                    ~chunk_id ~fragment_id
+                                                    ~location -> ())
+                        ~nsm_host_connection_pool_size
+                        ~osd_connection_pool_size
+                        ~osd_timeout
+                        ~default_osd_priority:Osd.High
+                        ~tls_config
+                        ~tcp_keepalive
+                        ~use_fadvise:true
+  in
+  let client = new Alba_client.alba_client base_client in
+  let make_object_name ~bid ~name =
+    serialize
+      (Llio.pair_to
+         Llio.int32_to
+         Llio.string_to)
+      (bid, name)
+  in
+  let with_namespace bid f =
+    match bucket_strategy with
+    | OneOnOne { prefix;
+                 preset; } ->
+       let namespace = prefix ^ (serialize ~buf_size:4 Llio.int32_to bid) in
+       Lwt.catch
+         (fun () -> f namespace)
+         (fun exn ->
+          client # create_namespace ~namespace ~preset_name:(Some preset) () >>= fun _ ->
+          f namespace)
+  in
+  object(self :# Fragment_cache.cache)
+    method add bid name blob =
+      Lwt.catch
+        (fun () ->
+         with_namespace
+           bid
+           (fun namespace ->
+            client # upload_object_from_bytes
+                   ~namespace
+                   ~object_name:(make_object_name ~bid ~name)
+                   ~object_data:blob
+                   ~checksum_o:None
+                   ~allow_overwrite:Nsm_model.Unconditionally >>= fun _ ->
+            Lwt.return_unit))
+        (fun exn ->
+         Lwt_log.debug_f ~exn "Exception while adding object to alba fragment cache")
+
+    method lookup bid name =
+      Lwt.catch
+        (fun () ->
+         with_namespace
+           bid
+           (fun namespace ->
+            client # download_object_to_bytes
+                   ~object_name:(make_object_name ~bid ~name)
+                   ~namespace
+                   ~consistent_read:false
+                   ~should_cache:true))
+        (fun exn ->
+         Lwt_log.debug_f ~exn "Exception during alba fragment cache lookup" >>= fun () ->
+         Lwt.return_none)
+
+    method lookup2 bid name object_slices =
+      Lwt.catch
+        (fun () ->
+         with_namespace
+           bid
+           (fun namespace ->
+            client # nsm_host_access # with_namespace_id
+                   ~namespace
+                   (fun namespace_id ->
+                    base_client # download_object_slices'
+                                ~namespace_id
+                                ~object_name:(make_object_name ~bid ~name)
+                                ~object_slices:(List.map
```

Comment:
anonymous function 
as a parameter in an anonymous function 
as a parameter in an anonymous function
as a parameter in an anonymous function.